### PR TITLE
add datadog_downtime module (try 3)

### PR DESCRIPTION
**Pull request by stefanweller**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

datadog_downtime
##### ANSIBLE VERSION

```
2.0.2.0
```
##### SUMMARY

Third try to bring datadog_downtime module to Ansible. Now with improved idempotency checking.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @stefanweller for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by stefanweller** [comment]
@gregdek Sorry, but had to remove the other fork due to rebasing problems.

**Comment by stefanweller** [comment]
@skornehl Please review again. THX

**Comment by stefanweller** [comment]
ready_for_review
works_for_me
passes_guidelines

**Comment by skornehl** [comment]
shipit

**Comment by goekhanm** [comment]
Awesome @stefanweller THX!! 
shipit !!
Our Opera-Team will love this module!

**Comment by skornehl** [comment]
@gregdek Does Ansibull-bot ignore this PR or is anything missing?

**Comment by gregdek** [comment]
The bot doesn't pick up new modules; they need to be approved manually. Putting this on the next new modules agenda.

File: monitoring/datadog_downtime.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    end = module.params['end']
+    message = module.params['message']
+    recurrence = module.params['recurrence']
+    loaded_recurrence = None
+
+    for downtime in downtimes:
+#        print downtime
+        # all scope tags must match - also the number of tags
+        if len(set(scope) - set(downtime['scope'])) == 0:
+            mismatch = False
```
Comment (alikins) [inline monitoring/datadog_downtime.py:197]: mismatch appears unused

File: monitoring/datadog_downtime.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+short_description: Set/update/cancel active or future downtimes (with optional recurrence) of DataDog monitors by a scope.
+description:
+- "Allows to set/update/cancel active or future downtimes (with optional recurrence) of DataDog monitors (www.datadoghq.com) by providing a scope."
+- "Uses U(http://docs.datadoghq.com/api/#monitors) API."
+- "Uses datadogpy U(https://github.com/DataDog/datadogpy) library -> pip install datadog."
+- "Currently this module has no check mode."
+version_added: "2.2.0"
+author: 'Stefan Weller <stefan.weller@asideas.de>, Github: stefanweller'
+requirements: [datadog]
+options:
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:33]: Hi, the formatting of the documentation isn't quite right here, could you please look at an existing module for examples of how to do this

File: monitoring/datadog_downtime.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    I(api_key):
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:34]: just `api_key`

File: monitoring/datadog_downtime.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            #end must match
+            if end and int(end) != downtime['end']:
+                continue
+
+            #all criteria match so no further action needed
+            module.exit_json(changed=False, msg="Matching downtime already present." % scope)
+
+    # set current time for start if no start given
+    if not start:
+        start = int(time.time())
```
Comment (alikins) [inline monitoring/datadog_downtime.py:236]: the 'time' module is available via the 

``` python
from ansible.module_utils.basic import *
```

but adding a 'import time' is recommended.

File: monitoring/datadog_downtime.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+- "Allows to set/update/cancel active or future downtimes (with optional recurrence) of DataDog monitors (www.datadoghq.com) by providing a scope."
+- "Uses U(http://docs.datadoghq.com/api/#monitors) API."
+- "Uses datadogpy U(https://github.com/DataDog/datadogpy) library -> pip install datadog."
+- "Currently this module has no check mode."
+version_added: "2.2.0"
+author: 'Stefan Weller <stefan.weller@asideas.de>, Github: stefanweller'
+requirements: [datadog]
+options:
+    I(api_key):
+        description: ["DataDog API key."]
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:35]: No need for `["` or `"]`
Have a look at existing modules for how to do this.

File: monitoring/datadog_downtime.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        required: true
+        default: null
+    I(app_key):
+        description: ["DataDog application key."]
+        required: true
+        default: null
+    I(state):
+        description: ["Designated state of the downtime. Currently 'present', 'updated' and 'absent' are supported."]
+        required: true
+        default: null
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:45]: Please add 
`choices:` and list present', 'updated' and 'absent', 

again, see other modules for examples

File: monitoring/datadog_downtime.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    I(api_key):
+        description: ["DataDog API key."]
+        required: true
+        default: null
+    I(app_key):
+        description: ["DataDog application key."]
+        required: true
+        default: null
+    I(state):
+        description: ["Designated state of the downtime. Currently 'present', 'updated' and 'absent' are supported."]
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:43]: Please remove the sentence
`Currently 'present', 'updated' and 'absent' are supported.`

When you add in `choices:`  the docs generator will add it into the table, e.g. http://docs.ansible.com/ansible/debconf_module.html

File: monitoring/datadog_downtime.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        required: False
+        default: null
+    I(message):
+        description: ["A downtime info text."]
+        required: False
+        default: null
+    I(active_only):
+        description: ["A C(True)/C(False) flag indicating that only active downtimes are affected or, if C(False), future ones too."]
+        required: False
+        default: null
```
Comment (gundalow) [inline monitoring/datadog_downtime.py:65]: Please add `choices:`

**Comment by gundalow** [comment]
needs_revision - docs updates

**Comment by alikins** [comment]
misc flake8 stuff (nothing blocking a merge, but for reference)

```
monitoring/datadog_downtime.py:139:0: F401 'sys' imported but unused
monitoring/datadog_downtime.py:149:0: E302 expected 2 blank lines, found 1
monitoring/datadog_downtime.py:185:0: E302 expected 2 blank lines, found 1
monitoring/datadog_downtime.py:247:0: E302 expected 2 blank lines, found 1
monitoring/datadog_downtime.py:274:0: E302 expected 2 blank lines, found 1
monitoring/datadog_downtime.py:283:0: E302 expected 2 blank lines, found 1
monitoring/datadog_downtime.py:194:0: E115 expected an indented block (comment)
monitoring/datadog_downtime.py:227:12: E265 block comment should start with '# '
monitoring/datadog_downtime.py:231:12: E265 block comment should start with '# '
monitoring/datadog_downtime.py:197:12: F841 local variable 'mismatch' is assigned to but never used
monitoring/datadog_downtime.py:236:20: F405 time may be undefined, or defined from star imports: ansible.module_utils.basic, ansible.module_utils.urls
monitoring/datadog_downtime.py:255:23: F405 time may be undefined, or defined from star imports: ansible.module_utils.basic, ansible.module_utils.urls
monitoring/datadog_downtime.py:212:73: E203 whitespace before ':'
monitoring/datadog_downtime.py:200:81: E203 whitespace before ':'
monitoring/datadog_downtime.py:204:89: E203 whitespace before ':'
monitoring/datadog_downtime.py:208:101: E203 whitespace before ':'
```

File: monitoring/datadog_downtime.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+def set_downtime(module, downtimes):
+    scope = module.params['scope']
+    start = module.params['start']
+    end = module.params['end']
+    message = module.params['message']
+    recurrence = module.params['recurrence']
+    loaded_recurrence = None
+
+    for downtime in downtimes:
+#        print downtime
```
Comment (mattclay) [inline monitoring/datadog_downtime.py:194]: This comment should be indented to match the block it is in.

**Pull request closed without merging**